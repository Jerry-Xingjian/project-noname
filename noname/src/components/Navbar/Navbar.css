.navbar-content-padding{
    padding: 0 15rem !important;
}

.navbar-brand {
    width: 25%;
}

.brand-name {
    margin: 0;
    width: 120px;
    font-size: 1.5rem;
    font-family: 'Rubik Dirt', cursive;
    font-weight: 700;
    background: -webkit-linear-gradient(left,#CF250D, #FFAE00);
    -webkit-background-clip: text;
    -webkit-text-fill-color: transparent;
    text-transform: uppercase;
    letter-spacing: 0.1em;
    text-decoration: none;
    transition: all 0.3s ease-in-out;
}

.menu-icon {
    width: 25%;
}

.menu-icon img {
    width: 1.5rem;
    height: 1.5rem;
    display: inline-block;
    color: #F8F9FA;
    transition: all 0.3s ease-in-out;
}