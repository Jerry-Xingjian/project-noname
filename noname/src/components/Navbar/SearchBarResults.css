.search-bar-results {
    height: 350px;
    overflow-y: auto;
    overflow-x: hidden;
}

.search-bar-results .username {
    line-height: 42px;
    margin-left: 8px !important;
}   

.user-result-li {
    padding: 0 1rem;
    cursor: pointer;
    transition: all 0.2s ease-in-out;
}

.user-result-li:hover {
    background-color: #F4F4F4;
}