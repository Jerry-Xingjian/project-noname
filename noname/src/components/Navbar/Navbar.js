import React, { useRef, useState } from 'react';
import jwtDecode from 'jwt-decode';
import { Link, useLocation } from 'react-router-dom';
import './Navbar.css';
import { localGet } from '../../utils/localStorage';
import { getActivityFeed } from '../../utils/api/activityFeed';
import { logout, searchUserByUsername } from '../../utils/api/user';
import SearchBarResults from './SearchBarResults';
import NewPostPopup from '../NewPostPopup/NewPostPopup';

function Navbar() {
  const token = localGet('token');
  let userId;
  let authFlag = false;
  if (token) {
    authFlag = true;
    const decoded = jwtDecode(token);
    userId = decoded.id;
  } else {
    authFlag = false;
  }
  const [showGlobalNewPostPopup, setShowGlobalNewPostPopup] = useState(false);
  const location = useLocation();
  const hideMenu = useRef(false);
  const [searchResults, setSearchResults] = useState([]);
  const [showResults, setShowResults] = useState(false);
  hideMenu.current = location.pathname === '/login' || location.pathname === '/register';

  const fetchPostData = async () => {
    await getActivityFeed();
    // setPosts(feed.data);
    // console.log(posts);
  };

  const inputHandler = async (e) => {
    // convert input text to lower case
    const { value } = e.target;
    if (value === '') {
      setSearchResults([]);
      setShowResults(false);
      return;
    }
    const res = await searchUserByUsername(value);
    setSearchResults(res.data.data);
    setShowResults(true);
  };

  const handleLogout = async () => {
    await logout();
  };

  return (
    <div className="fixed-top">
      {
          authFlag === true
            ? (
              <div>
                <nav className="navbar navbar-expand-lg bg-dark">
                  <div className="container-fluid navbar-content-padding d-flex justify-content-between">
                    <Link className="navbar-brand" to="/home">
                      <p className="brand-name">NoName</p>
                    </Link>

                    {/* Search bar */}
                    <input className={`form-control me-2 w-25 ${hideMenu.current ? 'd-none' : ''}`} type="search" placeholder="Search" aria-label="Search" onChange={inputHandler} />

                    <div className={`menu-icon d-flex justify-content-around ${hideMenu.current ? 'd-none' : ''}`}>
                      {/* Add route links here on each icon */}
                      <Link className="btn btn-link btn-sm" id="home-page-btn" to="/home">
                        <img src="https://nonome-project-media.s3.amazonaws.com/icons/global.svg" alt="menu" />
                      </Link>
                      <button type="button" className="btn btn-link btn-sm" onClick={() => setShowGlobalNewPostPopup(true)}>
                        <img src="https://nonome-project-media.s3.amazonaws.com/icons/add-square.svg" alt="menu" />
                      </button>
                      <Link className="btn btn-link btn-sm" to={`/profile/${userId}`}>
                        <img src="https://nonome-project-media.s3.amazonaws.com/icons/profile-circle.svg" alt="menu" />
                      </Link>
                      <Link onClick={handleLogout} className="btn btn-link btn-sm" to="/login">
                        <img src="https://nonome-project-media.s3.amazonaws.com/icons/logout.svg" alt="menu" />
                      </Link>
                    </div>
                  </div>
                </nav>
                <SearchBarResults className={`d-flex justify-content-center ${hideMenu.current ? 'd-none' : ''}`} searchResults={searchResults} showList={showResults} />
                <NewPostPopup
                  show={showGlobalNewPostPopup}
                  onHide={() => setShowGlobalNewPostPopup(false)}
                  userId={`${userId}`}
                  fetchPosts={fetchPostData}
                />
              </div>
            )
            : (
              <nav className="navbar navbar-expand-lg bg-dark">
                <div className="container-fluid navbar-content-padding d-flex justify-content-between">
                  <p className="brand-name">NoName</p>
                </div>
              </nav>
            )
      }
    </div>
  );
}

export default Navbar;
