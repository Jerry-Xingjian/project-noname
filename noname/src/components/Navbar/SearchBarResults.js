/* eslint-disable no-underscore-dangle */
import React from 'react';
import { Link } from 'react-router-dom';
// import axios from 'axios';
import './SearchBarResults.css';

function SearchBarResults(props) {
  const { searchResults, showList } = props;
  // const redirectToProfile = () => {
  //   console.log('clicked');
  //   redirect('/login');
  // };
  return (
    <div className={`search-bar-results w-100 d-flex justify-content-center ${searchResults.length === 0 ? 'd-none' : ''}`}>
      <ul className="list-group w-25">
        {
          searchResults.length > 0 ? (
            searchResults.map((item) => (
              <Link key={item._id} className="text-decoration-none" to={`/profile/${item._id}`}>
                <li className="list-group-item user-result-li">
                  <div className="d-flex">
                    <img
                      className="avatar-image ms-3 my-2"
                      src={item.profilePicture}
                      alt="Avatar"
                      mode="aspectFit"
                    />
                    <h6 className="m-0 username">{item.username}</h6>
                  </div>
                </li>
              </Link>
            ))
          ) : (
            <li className={`list-group-item user-result-li ${showList ? '' : 'd-none'}`}>
              <div className="d-flex">
                <h6 className="m-0 username">No results found</h6>
              </div>
            </li>
          )
        }
      </ul>
    </div>
  );
}

export default SearchBarResults;
