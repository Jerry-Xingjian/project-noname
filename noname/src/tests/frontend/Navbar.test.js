import { render, screen, fireEvent } from "@testing-library/react";
import renderer from 'react-test-renderer';
import jwtDecode from "jwt-decode";
import {createMemoryHistory} from 'history';
import React from "react";
import {Router} from 'react-router-dom'
import '@testing-library/jest-dom'
import Navbar from '../../components/Navbar/Navbar';

jest.mock('jwt-decode', () => jest.fn());

describe("snapshot test", () => {
  test('snapshot', () => {
		jwtDecode.mockImplementationOnce(() => ({ id: "6388320305d016d46329331d" }));
		const history = createMemoryHistory();
    const component = renderer.create(
			<Router location={history.location} navigator={history}>
				<Navbar />
			</Router>);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
})

describe("getByText test", () => {
  test('renders logo name', () => {
    const history = createMemoryHistory();
		jwtDecode.mockImplementationOnce(() => ({ id: "6388320305d016d46329331d" }));
    const { getByText } = render(
      <Router location={history.location} navigator={history}>
        <Navbar />
      </Router>,
    );
    const linkElement = getByText(/NoName/);
    expect(linkElement).toBeInTheDocument();
  });
})

// describe("Event test", () => {
//   test("Search input text change test", async () => {
//     const history = createMemoryHistory();
//     render(
//       <Router location={history.location} navigator={history}>
//         <Navbar />
//       </Router>
//     )
//     const search = screen.getByPlaceholderText("Search");
//     fireEvent.change(search, {target: {value: 'zy'}})
//     expect(search).toHaveValue('zy')
//   });
// });



// describe("Navigation test", () => {
//   test("Register navigate to login page", async () => {
//     // test for different effects than the page actually changing
//     const history = createMemoryHistory();
//     const spy = jest.spyOn(history, 'push');
//     render(
//       <Router location={history.location} navigator={history}>
//         <RegisterTable />
//       </Router>,
//     )
//     const user = userEvent.setup()
//     expect(screen.getByText(/Login here!/i)).toBeInTheDocument()
//     await user.click(screen.getByText(/Login here!/i))

//     // check that the content changed to the new page
//     expect(history.location.pathname).toBe('/login');
//     expect(spy).toBeCalled();
//   });

//   test("Login navigate to home page", async () => {
//     // test for different effects than the page actually changing
//     const history = createMemoryHistory();
//     const onSubmit = jest.fn((e) => {e.preventDefault()})
//     render(
//       <Router location={history.location} navigator={history}>
//         <RegisterTable handleSubmit = {onSubmit}/>
//       </Router>
//     )
//     const user = userEvent.setup()
//     // the belowed commented code is not complete, need improved
//     // await fireEvent.click(screen.getByRole('button', { name: 'Sign In' }), {
//     //     target: {value : 'xwang7@gmail.com'}
//     // });
//     // expect(history.location.pathname).toBe('/home');
//     // expect(onSubmit).toHaveBeenCalledTimes(1);
//   });
// });

// describe("Snapshot test", () => {
//   test('RegisterTable component matches snapshot', () => {
//     const history = createMemoryHistory();
//     const component = renderer.create(
//       <Router location={history.location} navigator={history}>
//         <RegisterTable />
//       </Router>
//     );
//     const tree = component.toJSON();
//     expect(tree).toMatchSnapshot();
//   });
// });