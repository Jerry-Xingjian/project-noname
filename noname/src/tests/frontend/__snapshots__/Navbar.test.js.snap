// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot test snapshot 1`] = `
<div
  className="fixed-top"
>
  <nav
    className="navbar navbar-expand-lg bg-dark"
  >
    <div
      className="container-fluid navbar-content-padding d-flex justify-content-between"
    >
      <a
        className="navbar-brand"
        href="/home"
        onClick={[Function]}
      >
        <p
          className="brand-name"
        >
          NoName
        </p>
      </a>
      <input
        aria-label="Search"
        className="form-control me-2 w-25 "
        onChange={[Function]}
        placeholder="Search"
        type="search"
      />
      <div
        className="menu-icon d-flex justify-content-around "
      >
        <a
          className="btn btn-link btn-sm"
          href="/home"
          onClick={[Function]}
        >
          <img
            alt="menu"
            src="https://nonome-project-media.s3.amazonaws.com/icons/global.svg"
          />
        </a>
        <button
          className="btn btn-link btn-sm"
          onClick={[Function]}
          type="button"
        >
          <img
            alt="menu"
            src="https://nonome-project-media.s3.amazonaws.com/icons/add-square.svg"
          />
        </button>
        <a
          className="btn btn-link btn-sm"
          href="/profile/6388320305d016d46329331d"
          onClick={[Function]}
        >
          <img
            alt="menu"
            src="https://nonome-project-media.s3.amazonaws.com/icons/profile-circle.svg"
          />
        </a>
      </div>
    </div>
  </nav>
  <div
    className="search-bar-results w-100 d-flex justify-content-center d-none"
  >
    <ul
      className="list-group w-25"
    />
  </div>
</div>
`;
